packages/core/src/validators.ts

```diff
@@ -1033,12 +1050,21 @@ export class IsNullOrUndefinedValidator extends Validator {
 
 export class NotEmptyValidator extends Validator {
   validatePath(value: any, path: Path, ctx: ValidationContext): PromiseLike<ValidationResult> {
-    return !isNullOrUndefined(value) && isNumber(value.length) && value.length > 0
-      ? ctx.successPromise(value)
-      : ctx.failurePromise(defaultViolations.notEmpty(path), value);
+    if (isNullOrUndefined(value) || (isObject(value) && !Object.keys(value).length) || (hasNumericLength(value) && !value.length)) {
+      return ctx.failurePromise(defaultViolations.notEmpty(path, value), value);
+    }
+    return ctx.successPromise(value);
   }
 }
 
+function isObject(value: any) {
```

Comment:
These methods (isObject and hasNumericLength) can be optimized by combining them directly to the NotEmptyValidator: there's no need to recheck if value is not null and `isArray(value)` doesn't need to run twice. These methods are also only used in that validator.